relay-server/src/services/buffer/envelope_buffer/mod.rs

```diff
@@ -204,155 +43,147 @@ struct EnvelopeBuffer<P: StackProvider> {
-                key: stack_key,
-                value: stack,
-            },
-            Priority {
-                readiness,
-                next_project_fetch,
-                ..
-            },
-        )) = self.priority_queue.peek_mut()
-        else {
-            return Ok(Peek::Empty);
-        };
+        relay_statsd::metric!(timer(RelayTimers::BufferPeek), {
+            let Some((&project_key_pair, priority)) = self.priority_queue.peek() else {
+                return Ok(Peek::Empty);
+            };
 
-        let ready = readiness.ready();
+            let envelope = self.envelope_repository.peek(project_key_pair).await?;
 
-        Ok(match (stack.peek().await?, ready) {
-            (None, _) => Peek::Empty,
-            (Some(envelope), true) => Peek::Ready(envelope),
-            (Some(envelope), false) => Peek::NotReady(*stack_key, *next_project_fetch, envelope),
+            Ok(match (envelope, priority.readiness.ready()) {
+                (None, _) => Peek::Empty,
+                (Some(envelope), true) => Peek::Ready(envelope),
+                (Some(envelope), false) => {
+                    Peek::NotReady(project_key_pair, priority.next_project_fetch, envelope)
+                }
+            })
         })
     }
 
     /// Returns the next-in-line envelope, if one exists.
     ///
-    /// The priority of the envelope's stack is updated with the next envelope's received_at
-    /// time. If the stack is empty after popping, it is removed from the priority queue.
+    /// The priority of the [`ProjectKeyPair`] is updated with the next envelope's received_at
+    /// time.
     pub async fn pop(&mut self) -> Result<Option<Box<Envelope>>, EnvelopeBufferError> {
-        let Some((QueueItem { key, value: stack }, _)) = self.priority_queue.peek_mut() else {
-            return Ok(None);
-        };
-        let project_key_pair = *key;
-        let envelope = stack.pop().await.unwrap().expect("found an empty stack");
-
-        let next_received_at = stack
-            .peek()
-            .await?
-            .map(|next_envelope| next_envelope.meta().start_time().into());
-
-        match next_received_at {
-            None => {
-                relay_statsd::metric!(counter(RelayCounters::BufferEnvelopeStacksPopped) += 1);
-                self.pop_stack(project_key_pair);
-            }
-            Some(next_received_at) => {
-                self.priority_queue
-                    .change_priority_by(&project_key_pair, |prio| {
-                        prio.received_at = next_received_at;
-                    });
+        relay_statsd::metric!(timer(RelayTimers::BufferPop), {
+            let Some((&project_key_pair, _)) = self.priority_queue.peek() else {
+                return Ok(None);
+            };
+
+            // There must be an envelope when popping, since we assume that `peek` is called before
+            // the call to `pop` and no concurrent access is performed in the meanwhile.
```

Comment:
```suggestion
            // There must be an envelope when popping, since we call `peek` in the statement above
            // and no concurrent access is performed in the meanwhile.
```